pkg/controller/cronjob/cronjob_controller2.go

```diff
@@ -0,0 +1,580 @@
+			recorder.Eventf(cj, v1.EventTypeWarning, "UnexpectedJob", "Saw a job that the controller did not create or forgot: %s", j.Name)
+			// We found an unfinished job that has us as the parent, but it is not in our Active list.
+			// This could happen if we crashed right after creating the Job and before updating the status,
+			// or if our jobs list is newer than our cj status after a relist, or if someone intentionally created
+			// a job that they wanted us to adopt.
+
+			// TODO: maybe handle the adoption case?  Concurrency/suspend rules will not apply in that case, obviously, since we can't
+			// stop users from creating jobs if they have permission.  It is assumed that if a
+			// user has permission to create a job within a namespace, then they have permission to make any cronJob
+			// in the same namespace "adopt" that job.  ReplicaSets and their Pods work the same way.
+			// TBS: how to update cj.Status.LastScheduleTime if the adopted job is newer than any we knew about?
+		} else if found && IsJobFinished(&j) {
+			_, status := getFinishedStatus(&j)
+			deleteFromActiveList(cj, j.ObjectMeta.UID)
+			recorder.Eventf(cj, v1.EventTypeNormal, "SawCompletedJob", "Saw completed job: %s, status: %v", j.Name, status)
+		}
+	}
+
+	// Remove any job reference from the active list if the corresponding job does not exist any more.
+	// Otherwise, the cronjob may be stuck in active mode forever even though there is no matching
+	// job running.
+	for _, j := range cj.Status.Active {
+		if found := childrenJobs[j.UID]; !found {
+			recorder.Eventf(cj, v1.EventTypeNormal, "MissingJob", "Active job went missing: %v", j.Name)
+			deleteFromActiveList(cj, j.UID)
+		}
+	}
+
+	// TODO: @alpatel explore if cached client can be used as realCJControl
+	updatedCJ, err := cjc.UpdateStatus(cj)
+	if err != nil {
+		klog.V(2).Infof("Unable to update status for %s (rv = %s): %v", nameForLog, cj.ResourceVersion, err)
+		return err, nil
+	}
+	*cj = *updatedCJ
+
+	if cj.DeletionTimestamp != nil {
+		// The CronJob is being deleted.
+		// Don't do anything other than updating status.
+		return fmt.Errorf("cronjob %s/%s is being deleted", cj.Namespace, cj.Name), nil
+	}
+
+	if cj.Spec.Suspend != nil && *cj.Spec.Suspend {
+		klog.V(4).Infof("Not starting job for %s because it is suspended", nameForLog)
+		return nil, nil
+	}
+
+	// TODO: alpatel: this is here now just to pass the unit tests,
+	//  move this to the start of the function to return on error early
+	sched, err := cron.ParseStandard(cj.Spec.Schedule)
+	if err != nil {
+		// this is likely a user error in defining the spec value
+		// we should log the error and not reconcile this cronjob until an update to spec
+		klog.Errorf("unparseable schedule: %s : %s", cj.Spec.Schedule, err)
+		recorder.Eventf(cj, v1.EventTypeWarning, "UnparseableSchedule", "unparseable schedule: %s : %s", cj.Spec.Schedule, err)
+		return nil, nil
+	}
+	times, err := getRecentUnmetScheduleTimes2(*cj, now, sched)
```

Comment:
Reading through this more, I don't think we need the separate method for this calculation.

Reply:
addressed this, PTAL, thanks